### PR TITLE
👌 IMP: Hogwild, prefetch and 150 super batches

### PR DESCRIPTION
**Pull request by ianagbip1oti**
```
-------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 7.96 +/- 5.05, nElo: 15.02 +/- 9.53
LOS: 99.90 %, DrawRatio: 51.43 %, PairsRatio: 1.18
Games: 5110, Wins: 1151, Losses: 1034, Draws: 2925, Points: 2613.5 (51.14 %)
Ptnml(0-2): [26, 543, 1314, 632, 40], WL/DD Ratio: 0.50
LLR: 2.92 (100.9%) (-2.25, 2.89) [0.00, 3.00]
--------------------------------------------------

--------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, DFRC_4852_v1.epd):
Elo: 11.47 +/- 7.15, nElo: 19.76 +/- 12.31
LOS: 99.92 %, DrawRatio: 46.73 %, PairsRatio: 1.25
Games: 3060, Wins: 758, Losses: 657, Draws: 1645, Points: 1580.5 (51.65 %)
Ptnml(0-2): [31, 332, 715, 409, 43], WL/DD Ratio: 0.58
LLR: 2.90 (100.4%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------

-------------------------------------------------
Results of princhess vs princhess-main (8+0.08, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 13.45 +/- 7.64, nElo: 23.03 +/- 13.07
LOS: 99.97 %, DrawRatio: 47.61 %, PairsRatio: 1.32
Games: 2714, Wins: 710, Losses: 605, Draws: 1399, Points: 1409.5 (51.93 %)
Ptnml(0-2): [34, 272, 646, 365, 40], WL/DD Ratio: 0.70
LLR: 2.90 (100.3%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------
```

## Summary by CodeRabbit

- **Improvements**
  - Training now processes twice as many super batches, supporting more extensive model training.
  - Training data is loaded and prefetched in the background to improve throughput.
  - CPU usage is better aligned with available physical processor capacity.
  - Gradient processing has been optimized for more efficient parallel training.
  - The first training checkpoint is now saved at the beginning of a run.

- **Updates**
  - Recorded model-training metadata now reflects the expanded 150-super-batch training configuration and improved loss value.

### CODE REVIEW

File: crates/princhess-train/src/bin/train-value.rs [hunk 1]
```diff
@@ -514,37 +516,42 @@ fn gradients_batch(
...
+    let gradients_ptr = gradients as *mut ValueNetwork as usize;
 
-    thread::scope(|s| {
+    let metrics: Vec<BatchMetrics> = thread::scope(|s| {
         batch
             .chunks(size)
             .map(|chunk| {
                 s.spawn(move || {
-                    let mut inner_gradients = ValueNetwork::zeroed();
+                    let grads = unsafe { &mut *(gradients_ptr as *mut ValueNetwork) };
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-value.rs:533]: This is intentional — it's the Hogwild algorithm. The whole point is to accept the data races as gradient noise rather than paying the cost of per-thread gradient copies and a serial reduction step. The previous
  code did exactly what you're suggesting and was our bottleneck: at 8+ threads the cost of allocating N×22.5 MB gradient buffers and reducing them dominated, causing throughput to degrade rather than scale.

  The SAFETY comment explains the reasoning: the feature weights are sparse (~32/9216 active per position), so concurrent writes to the same row are rare. The output layer is dense but tiny (640 floats), and
  empirically the gradient noise is acceptable for SGD — Hogwild is a well-studied technique for exactly this case (Recht et al. 2011). We ran 150 super-batches with this implementation and achieved a final loss of
  0.03528 vs 0.03593 for the previous per-thread approach.

[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-value.rs:533]: You're right that this is UB in Rust's memory model — multiple &mut ValueNetwork to the same allocation is undefined regardless of whether races actually occur. The SAFETY comment has been updated to state this
  explicitly.

  The alternatives we considered:
  - ptr::write / raw pointer writes directly: removes the multiple-&mut aliasing, but concurrent writes through raw pointers to the same memory are still UB in Rust's model — same problem, slightly different spelling
  - Atomic accumulation: AtomicF32 isn't stable, and a CAS loop over the ~5.9M gradient floats per batch would be slower than the per-thread reduction we removed
  - Exclusive write regions: the sparse feature weights are indexed by position features, not by thread, so partitioning requires either a preprocessing pass or a sharded gradient layout that significantly
    complicates the backprop path

  This is a training binary, not engine code, and we've accepted the unsoundness knowingly. LLVM won't miscompile this access pattern on x86-64 today, and the practical outcome (2× speedup, better final loss)
  validates the tradeoff.

[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-value.rs:533]: ok, noted as a valid issue. Will switch to ptr::write in a subsequent PR (not this one since training and testing is already well progressed)
